### PR TITLE
Add automated backup and restore tests

### PR DESCRIPTION
**Pull request by pcahyna**
This PR adds two Beakerlib ( https://github.com/beakerlib/beakerlib/wiki/man ) tests for the complete unattended backup and restore process:
- make-backup-and-restore-bios, by @lzaoral. This tests backup and restore using the USB method, backup is stored to a second disk drive, which is also used to boot the rescue system from. This test can not be executed automatically in CI until the test infrastructure we are using (Testing Farm) supports adding a second disk to the VMs where the tests are executing. It serves as a basis for the second test:
- make-backup-and-restore-iso, by @antonvoznia. This uses the ISO output. A ramdisk is used to emulate the CD-ROM drive. The ISO image with backup gets written to the ramdisk after ReaR recovery starts and before it starts reformatting the disk (where the ISO image is stored) to ensure that the backup persists until we need to restore it. The changes relative to make-backup-and-restore-bios are described in "A. Voznia: Automation of testing of operating system backup and recovery (2023)" https://dspace.cvut.cz/handle/10467/107238.

Solves one part of https://github.com/rear/rear/issues/3014 (tests will need to enabled in the rear repository in order to get executed in CI).

### CODE REVIEW

**Review by gdha — APPROVED** [review]

**Comment by gdha** [comment]
@pcahyna you should now have the required permissions to merge the PR

File: tests/make-backup-and-restore-iso/runtest.sh [hunk 1]
```diff
@@ -73,13 +73,15 @@ rlJournalStart
...
         rlPhaseStartSetup "Create /etc/rear/local.conf"
-            rlFileBackup "/etc/rear/local.conf"
+            rlFileBackup "$REAR_CONFIG"
             rlRun -l "echo 'OUTPUT=USB
 BACKUP=NETFS
 BACKUP_URL=usb://$REAR_LABEL_PATH
-USER_INPUT_TIMEOUT=10' | tee /etc/rear/local.conf" \
+USER_INPUT_TIMEOUT=10
+ISO_DEFAULT=automatic
+ISO_RECOVER_MODE=unattended' | tee $REAR_CONFIG" \
```
Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh:107]: These changes are included in 4cafe306d7e7c0f7e74dc90648295e0d05cb8987 but do not seem to be related to it at all.

File: tests/make-backup-and-restore-iso/runtest.sh [hunk 2]
```diff
@@ -73,13 +73,15 @@ rlJournalStart
         rlPhaseEnd
 
         rlPhaseStartSetup "Create /etc/rear/local.conf"
```
Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh]: You probably want to use `$REAR_CONFIG` here as well?

File: tests/make-backup-and-restore-iso/runtest.sh [hunk 3]
```diff
@@ -280,8 +282,8 @@ LABEL rear
 
         rlPhaseStartCleanup
             rlFileRestore
-            rlRun "rm -f drive_layout.{old,new}" 0 "Remove lsblk outputs"
-            rlRun "rm -rf /root/rear*.log /var/log/rear/*" 0 "Remove ReaR logs"
+            rlRun "rm -f $REAR_HOME_DIRECTORY/drive_layout.{old,new}" 0 "Remove lsblk outputs"
+            rlRun "rm -rf rear*.log /var/log/rear/*" 0 "Remove ReaR recovery log"
```
Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh]: Will the plain `rear*.log` glob remove correct files? Few lines above, the same file as referenced with the `$REAR_HOME_DIRECTORY` prefix.  (Also, the message change seems unnecessary.)

File: tests/make-backup-and-restore-iso/runtest.sh [hunk 4]
```diff
@@ -138,122 +107,14 @@ ISO_RECOVER_MODE=unattended' | tee $REAR_CONFIG" \
...
-            rlRun "cat /usr/share/syslinux/mbr.bin > /dev/$ROOT_DEVICE" \
-                0 "Write EXTLINUX to /dev/$ROOT_DEVICE MBR"
-
-            if ! rlGetPhaseState; then
-                rlDie "FATAL ERROR: Installing EXTLINUX failed"
-            fi
+            rlLog "Setup GRUB"
+            rlRun "echo 'menuentry \"ReaR-recover\" {
+  loopback loop (hd0,msdos2)/root/rear/var/lib/rear/output/rear-fedora.iso
+  linux (loop)/isolinux/kernel rw selinux=0 console=ttyS0,9600 console=tty0 auto_recover unattended
```
Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh]: Are you sure that we will always use these hard-coded `console` devices?  What was wrong with the detection implemented in the`rlPhaseStartSetup "Redirect ReaR output to $CONSOLE_DEVICE"` phase which was removed?

Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh]: Also, is it intended that we boot the recovery directly from GRUB?  I know that it has some issues with chainloading but it would be nice to also test the bootloader (ISOLINUX AFAIK) stored on the ISO image.

File: tests/make-backup-and-restore-iso/runtest.sh [hunk 5]
```diff
@@ -60,9 +60,12 @@ check_and_submit_rear_log() {
     rlFileSubmit $path "rear-$1.log"
 }
 
+ROOT_DISK=$(df -hT | grep /$ | awk '{print $1}')
```
Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh:70]: If this works reliably enough then fine, but device naming in linux is not stable.

edit: typo

File: tests/make-backup-and-restore-iso/runtest.sh [hunk 6]
```diff
@@ -107,12 +110,19 @@ ISO_RECOVER_MODE=unattended' | tee $REAR_CONFIG" \
             rlAssertExists recovery_will_remove_me
         rlPhaseEnd
 
+        rlPhaseStartSetup "Make small iso file that is bootable by memdisk"
+            rlRun "xorriso -as mkisofs -r -V 'REAR-ISO' -J -J -joliet-long -cache-inodes -b isolinux/isolinux.bin -c isolinux/boot.cat -boot-load-size 4 -boot-info-table -no-emul-boot -eltorito-alt-boot -dev $REAR_ISO_OUTPUT/rear-$HOSTNAME_SHORT.iso -o $REAR_ISO_OUTPUT/small-rear.iso -- -rm_r backup"
+        rlPhaseEnd
+
```
Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh:142]: Nitpick: two empty lines.

File: tests/make-backup-and-restore-iso/runtest.sh [hunk 7]
```diff
@@ -107,12 +110,19 @@ ISO_RECOVER_MODE=unattended' | tee $REAR_CONFIG" \
...
+
         rlPhaseStartSetup "Force the machine to autoboot the ReaR rescue system"
+            rlLog "Setup Boot"
+            rlLog "Copying memdisk"
+            rlRun "cp /usr/share/syslinux/memdisk /boot/"
```
Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh]: These commands can be combined because `rlRun` can specify comments.

File: tests/make-backup-and-restore-iso/runtest.sh [hunk 8]
```diff
@@ -138,122 +107,14 @@ ISO_RECOVER_MODE=unattended' | tee $REAR_CONFIG" \
...
-    APPEND hd1' | tee /boot/extlinux/extlinux.conf" \
-                0 "Save EXTLINUX configuration"
-            rlRun "cat /usr/share/syslinux/mbr.bin > /dev/$ROOT_DEVICE" \
-                0 "Write EXTLINUX to /dev/$ROOT_DEVICE MBR"
-
-            if ! rlGetPhaseState; then
-                rlDie "FATAL ERROR: Installing EXTLINUX failed"
-            fi
+            rlLog "Setup GRUB"
+            rlRun "echo 'menuentry \"ReaR-recover\" {
```
Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh]: These commands can be combined because `rlRun` can specify comments.

File: tests/make-backup-and-restore-iso/runtest.sh (hunk 4 above)
Comment (lzaoral) [reply tests/make-backup-and-restore-iso/runtest.sh]: Ok, it looks like next commits have done exactly that...

Comment (lzaoral) [reply tests/make-backup-and-restore-iso/runtest.sh]: It looks like next commits do not even need these parameters, right?

**Review by lzaoral — CHANGES_REQUESTED** [review]
I've tried the `make-backup-and-restore-iso` test locally and it works!

Nitpicks:
* Outdated years in copyright.
* You probably want to update authors in the remaining files in the `make-backup-and-restore-iso` directory.
* I'd squash the `make-backup-and-restore-iso` related commits before merging.

I'd add more comments what is the `make-backup-and-restore-iso` test actually trying to achieve.
While it is great that it works, I have no idea how and why are the `memdisk` shenanigans required.

**Comment by pcahyna** [comment]
@lzaoral I believe I addressed most of your comments. Please check whether the comments added to the script are explanatory enough.

File: tests/make-backup-and-restore-iso/runtest.sh (hunk 3 above)
Comment (pcahyna) [reply tests/make-backup-and-restore-iso/runtest.sh]: I believe now that `rear*.log` should be always referenced as `/root/rear*.log` and not as `$REAR_HOME_DIRECTORY/rear*.log`. So, few lines above it is wrong. This is because this log is created by ReaR recovery and it creates it under /root unconditionally (more precisely, under `$ROOT_HOME_DIR`, which is set to `~root` by default). So it is not affected by our `$REAR_HOME_DIRECTORY` setting.

Comment (pcahyna) [reply tests/make-backup-and-restore-iso/runtest.sh]: @lzaoral the last commit should resolve all the `$REAR_HOME_DIRECTORY` confusion ...

File: tests/make-backup-and-restore-iso/runtest.sh [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        rlPhaseStartTest "Assert that the recovery was successful"
+            rlAssertNotExists $REAR_HOME_DIRECTORY/recovery_will_remove_me
+
+            rlAssertExists $REAR_HOME_DIRECTORY/drive_layout.old
+            rlAssertExists /root/rear*.log
+
+            rlRun -l "lsblk | tee $REAR_HOME_DIRECTORY/drive_layout.new" \
+                0 "Get current lsblk output"
+            if ! rlAssertNotDiffer $REAR_HOME_DIRECTORY/drive_layout.old $REAR_HOME_DIRECTORY/drive_layout.new; then
+                rlRun -l "diff -u $REAR_HOME_DIRECTORY/drive_layout.old $REAR_HOME_DIRECTORY/drive_layout.new" \
```
Comment (lzaoral) [inline tests/make-backup-and-restore-iso/runtest.sh]: Small nitpick:
```suggestion
            if ! rlAssertNotDiffer $REAR_HOME_DIRECTORY/drive_layout.{old,new}; then
                rlRun -l "diff -u $REAR_HOME_DIRECTORY/drive_layout.{old,new}" \
```

**Review by lzaoral — CHANGES_REQUESTED** [review]
Thanks for the update, @pcahyna!  I was not able to successfully execute the test using `tmt` on Fedora Rawhide (it might be useful to also add a default test plan in the future) even after applying the following patch:
```diff
diff --git a/tests/make-backup-and-restore-iso/main.fmf b/tests/make-backup-and-restore-iso/main.fmf
index 286bb23..4db7762 100644
--- a/tests/make-backup-and-restore-iso/main.fmf
+++ b/tests/make-backup-and-restore-iso/main.fmf
@@ -1,6 +1,7 @@
 summary: 'Test basic functionality of ReaR on systems with BIOS using bootable ISO image'

 require:
+    - rear
     - syslinux-extlinux
     - syslinux-nonlinux
     - xorriso

```
The test timeouted after a while.  Direct execution though `1mt` also ended with a failure during boot of the rescue environment.   Very small nitpick below.  Also, I'd still squash the commits that lead to the final version of `make-backup-and-restire-iso` before merging.

**Comment by lzaoral** [comment]
As noted by @pcahyna in a private conversation, the test (or more probably ReaR) is not yet compatible with Fedora Rawhide.  I was able to successfully execute the test both using `tmt` and `1mt` on a RHEL 9 machine with a sufficiently large amount of RAM.

**Comment by pcahyna** [comment]
@lzaoral the test plan will be in the ReaR repository: https://github.com/pcahyna/rear/pull/13 . It also includes information about the excluded Fedora versions and the reason behind.

**Comment by pcahyna** [comment]
I squashed and reworded commits without any content changes. Merging.

**Comment by pcahyna** [comment]
> As noted by @pcahyna in a private conversation, the test (or more probably ReaR) is not yet compatible with Fedora Rawhide.

Solved in PR https://github.com/rear/rear/pull/3079
